src/lib/provable/gadgets/blake2b.ts

```diff
@@ -0,0 +1,282 @@
+          .add(buf[i * 8 + 7].value.mul(1n << 56n)).toFields()
+      )
+    );
+  }
+
+  for (let i = 0; i < 12; i++) {
+    // twelve rounds
+    const s = BLAKE2BConstants.SIGMA[i % 10];
+    G(v, 0, 4, 8, 12, m[s[0]], m[s[1]]);
+    G(v, 1, 5, 9, 13, m[s[2]], m[s[3]]);
+    G(v, 2, 6, 10, 14, m[s[4]], m[s[5]]);
+    G(v, 3, 7, 11, 15, m[s[6]], m[s[7]]);
+
+    G(v, 0, 5, 10, 15, m[s[8]], m[s[9]]);
+    G(v, 1, 6, 11, 12, m[s[10]], m[s[11]]);
+    G(v, 2, 7, 8, 13, m[s[12]], m[s[13]]);
+    G(v, 3, 4, 9, 14, m[s[14]], m[s[15]]);
+  }
+
+  for (let i = 0; i < 8; i++) {
+    // XOR the two halves
+    h[i] = v[i].xor(v[i + 8]).xor(h[i]);
+  }
+  return state;
+}
+
+/**
+ * Initializes the state with the given digest length.
+ *
+ * @param {number} outlen - Digest length in bits
+ * @returns {State}
+ */
+function initialize(outlen: number): {
+  h: UInt64[];
+  t: UInt64[];
+  buf: UInt8[];
+  buflen: number;
+  outlen: number;
+} {
+  const h = BLAKE2B.IV.slice(); // shallow copy IV to h
+  h[0] = UInt64.from(0x01010000).xor(UInt64.from(outlen)).xor(h[0]); // state "param block"
+
+  return {
+    h,
+    t: [UInt64.zero, UInt64.zero],
+    buf: [],
+    buflen: 0,
+    outlen,
+  };
+}
+
+/**
+ * Updates hash state
+ * @param {State} state
+ * @param {UInt8[]} input
+ * @returns {State} updated state
+ */
+function update(
+  state: {
+    h: UInt64[];
+    t: UInt64[];
+    buf: UInt8[];
+    buflen: number;
+    outlen: number;
+  },
+  input: UInt8[]
+): State {
+  for (let i = 0; i < input.length; i++) {
+    if (state.buflen === 128) {
+      // buffer full ?
+      state.t[0] = state.t[0].addMod64(UInt64.from(state.buflen)); // add counters
+      if (state.t[0].toBigInt() < state.buflen) {
+        // carry overflow ?
+        state.t[1] = state.t[1].addMod64(UInt64.one); // high word
+      }
+      state = compress(state, false); // compress (not last)
+      state.buflen = 0; // counter to zero
+    }
+    state.buf[state.buflen++] = input[i];
+  }
+  return state;
+}
+
+/**
+ * Finalizes the hash state and returns digest
+ * @param {State} state
+ * @returns {UInt8[]} digest
+ */
+function final(state: {
+  h: UInt64[];
+  t: UInt64[];
+  buf: UInt8[];
+  buflen: number;
+  outlen: number;
+}): UInt8[] {
+  state.t[0] = state.t[0].addMod64(UInt64.from(state.buflen)); // mark last block offset
+  if (state.t[0].toBigInt() < state.buflen) {
+    // carry overflow ?
+    state.t[1] = state.t[1].addMod64(UInt64.one); // high word
+  }
+  /*
+  state.t[1] = state.t[1].add(
+    Provable.if(
+      state.t[0].lessThan(UInt64.from(state.buflen)),
+      UInt64.one,
+      UInt64.zero
+    )
+  );
+*/
+  while (state.buflen < 128) {
+    state.buf[state.buflen++] = UInt8.from(0); // fill up with zeroes
+  }
+  compress(state, true);
+
+  // little endian convert and store
+  const out: UInt8[] = [];
+  for (let i = 0; i < state.outlen; i++) {
+    out[i] = UInt8.from(
+      state.h[i >> 3].rightShift(8 * (i & 7)).and(UInt64.from(0xff))
```

Comment:
if you aim for efficiency: here you can also use arithmetic instead of bitwise operations

Reply:
I tried to replace this line with `state.h[i >> 3].div(2 ** (8 * (i & 7))).mod(0xff)` but it increased the row number a lot. I am not sure if I'm missing something.

Reply:
You can check out the bytesToWord function that's used in the keccak and sha2 implementations!
Probably a good idea to use that function

Reply:
I meant `wordToBytes()` btw 😅

Reply:
yeah, I tried that, but it doesn’t work for this case. `wordToBytes()` takes a word as `Field` and converts it into an array of bytes as `UInt8[]`. Here, the logic is more complex. I don’t think a specific function in `bit-slices.ts` is necessary for this case since it would only be used here.

Reply:
The way I read this logic, it just converts each 64-bit state.h element to 8 bytes (UInt8), in little indian order. So it's the same as wordToBytes, you should be able to use something like
```ts
let out = state.h.slice(0, 8*outlen)
  .flatMap(x => wordToBytes(x.value))
```

Reply:
It works! Thanks.